### PR TITLE
Date locales

### PR DESCRIPTION
**Pull request by paulcpederson**
**Related Issue:** #585

## Summary

Main update is using the localization files from the localization team. Also fixes several smaller open issues regarding the date component.

- feat(date): add full localization support (#585, #979)
- fix(date): fire change event when editing year (#1017)
- feat(date): respect language year-month order in date-month-header (#908)
- feat(date): add year suffix functionality for ja, ko, ch (#907)
- fix(date): allow 3 digit years in calendar year input (#905)
- fix(date): only update proxy input on real updates (#896)
- fix(date): don't fire event on programatic value change (#722)

### CODE REVIEW

File: src/utils/date.ts [hunk 1]
```diff
@@ -94,3 +96,62 @@ export function nextMonth(date: Date): Date {
...
+      .filter((num) => num)
+      .join("")
+  );
+}
+
+/**
+ * Parse numeric units for day, month, and year from a localized string
+ * month starts at 0 (can pass to date constructor)
+ */
+export function parseDateString(str: string, localeData: DateLocaleData): { day: number; month: number; year: number } {
```
Comment (eriklharper) [inline src/utils/date.ts:128]: Might be nice to have the return type of the function be declared with a name like the `unitOrderSignifier` type down below.

**Comment by paulcpederson** [comment]
Weird the test runs locally, seems like it only fails on travis, and only sometimes?

File: src/components/calcite-date/calcite-date.tsx [hunk 2]
```diff
@@ -193,19 +212,68 @@ export class CalciteDate {
...
+  private async loadLocaleData(): Promise<void> {
+    const { locale, visible } = this;
+
+    if (!Build.isBrowser || !locale || !visible) {
+      return;
+    }
+    this.localeData = await getLocaleData(locale);
+    // we need locale data to parse initial date from value
+    if (this.value && !this.valueAsDate) {
+    }
```
Comment (eriklharper) [inline src/components/calcite-date/calcite-date.tsx:249]: ☝️ something missing here?

File: src/components/calcite-date-day/calcite-date-day.tsx [hunk 3]
```diff
@@ -86,11 +87,15 @@ export class CalciteDateDay {
   //
   //--------------------------------------------------------------------------
   render(): VNode {
-    const intl = new Intl.NumberFormat(this.locale);
+    // this.localeData.numerals.split("")
```
Comment (eriklharper) [inline src/components/calcite-date-day/calcite-date-day.tsx:90]: Should this commented line be removed? ☝️

File: src/components/calcite-date/calcite-date.tsx (hunk 2 above)
Comment (paulcpederson) [reply src/components/calcite-date/calcite-date.tsx:249]: good catch

**Comment by jcfranco** [comment]
@paulcpederson I hate to ask this, but can you split this up into multiple PRs? Reasoning:

1. Simplifies the review process
2. Keeps feedback/changes from PRs focused (this also helps blocking unrelated changes that are good to go)
3. Ensures changelog entries are informative since it avoids coalescing commit messages

**Comment by paulcpederson** [comment]
@jcfranco what's the best way to do that?

**Comment by jcfranco** [comment]
Not sure if there's a best way, but if you have commits with isolated changes, you could checkout `master`, create a feature/fix branch, cherry pick them. Otherwise, you may have to create a feature/fix branch on `date-locales`, soft reset to `master` and only commit relevant changes.

**Comment by paulcpederson** [comment]
Yeah it was a bit more complicated than that, though, as several of the issues interact. There were also a variety of merge conflicts from the function return types that were added. Cherry picking is not possible on a level more granular than a commit afaict.

I will try the soft reset idea and see if I can do this.

**Comment by paulcpederson** [comment]
OK, first and most important pr is here: https://github.com/Esri/calcite-components/pull/1043 

I think it will be easier to open the remainder of the prs on top of that, so I'll wait until that makes it into `master` to open the others.

**Pull request closed without merging**